lightning/src/routing/network_graph.rs

```diff
@@ -373,21 +373,40 @@ impl<C: Deref , L: Deref > RoutingMessageHandler for NetGraphMsgHandler<C, L> wh
 
 		let mut pending_events = self.pending_events.lock().unwrap();
 		let batch_count = batches.len();
+		let mut prev_batch_endblock = msg.first_blocknum;
 		for (batch_index, batch) in batches.into_iter().enumerate() {
-			// Per spec, the initial first_blocknum needs to be <= the query's first_blocknum and subsequent
-			// must be >= the prior reply. We'll simplify this by using zero since its still spec compliant and
-			// sequence completion is now explicitly.
-			let first_blocknum = 0;
-
-			// Per spec, the final end_blocknum needs to be >= the query's end_blocknum, so we'll use the
-			// query's value. Prior batches must use the number of blocks that fit into the message. We'll
-			// base this off the last SCID in the batch since we've somewhat abusing first_blocknum.
+			// Per spec, the initial `first_blocknum` needs to be <= the query's `first_blocknum`
+			// and subsequent `first_blocknum`s must be >= the prior reply's `first_blocknum`.
+			//
+			// Additionally, c-lightning versions < 0.10 require that the `first_blocknum of each
+			// reply is >= the previous reply's `first_blocknum` and either exactly the previous
+			// reply's `first_blocknum + number_of_blocks` or exactly one greater. This is a
+			// significant diversion from the requirements set by the spec, and, in case of blocks
+			// with no channel opens (eg empty blocks), requires that we use the previous value and
+			// *not* derive the first_blocknum from the actual first block of the reply.
+			let first_blocknum = prev_batch_endblock;
+
+			// Each message carries the number of blocks (from the `first_blocknum`) its contents
+			// fit in. Though there is no requirement that we use exactly the number of blocks its
+			// contents are from, except for the bogus requirements c-lightning enforces, above.
+			//
+			// Per spec, the last end_block needs to be >= the query's `end_blocknum`.
+			// Thus, for the last reply, we calculate the difference between the query's
+			// `end_blocknum` and the start of the reply.
```

Comment:
Earlier you use `first_blocknum + number_of_blocks` for `end_blocknum`. Given `end_blocknum` is not used in the spec, should we be consistent and expand it out here? Prior to the change, we were consistently using `end_blocknum` throughout.

Reply:
I went for defining the "end block" and then using it as a non-` term.

```
-                       // Per spec, the last end_block needs to be >= the query's `end_blocknum`.
-                       // Thus, for the last reply, we calculate the difference between the query's
-                       // `end_blocknum` and the start of the reply.
+                       // Per spec, the last end block (ie `first_blocknum + number_of_blocks`) needs to be
+                       // >= the query's end block. Thus, for the last reply, we calculate the difference
+                       // between the query's end block and the start of the reply.
                        //
```